### PR TITLE
End-to-end publishEntry tests

### PR DESCRIPTION
**Pull request by sandreae**
End-to-end tests (including many failing cases) for publishing entries.

~This PR throws up a few `TODO`s as I find some panicking cases. I suggest not fixing them in this PR but instead raising issues.~

~Probably worth waiting for this: https://github.com/p2panda/p2panda/pull/367 then we can introduce tests for these new errors and more invalid encoded entries as well.~

Targeting p2panda-rs branch which fixes `EntrySigned` validation: https://github.com/p2panda/p2panda/tree/improve-validation-in-entry-signed

Follow-up issues: https://github.com/p2panda/aquadoggo/issues/160 https://github.com/p2panda/aquadoggo/issues/159

- [x] BEFORE MERGE: merge this in `p2panda-rs` and then switch target branch back to `main`

## 📋 Checklist

- [x] Add tests that cover your changes
- [x] Add this PR to the _Unreleased_ section in `CHANGELOG.md`
- [x] Link this PR to any issues it closes
- [x] New files contain a SPDX license header

### CODE REVIEW

File: aquadoggo/src/graphql/client/mutation.rs [hunk 1]
```diff
@@ -110,21 +124,34 @@ mod tests {
...
 
+    const CREATE_OPERATION_WITH_PREVIOUS_OPS: &str = "A566616374696F6E6663726561746566736368656D617849636861745F30303230633635353637616533376566656132393365333461396337643133663866326266323364626463336235633762396162343632393331313163343866633738626776657273696F6E017370726576696F75735F6F7065726174696F6E738178443030323036356637346636666438316562316261653139656230643864636531343566616136613536643762343037366437666261343338353431303630396232626165666669656C6473A1676D657373616765A26474797065637374726576616C75657357686963682049206E6F77207570646174652E";
+
+    const DELETE_OPERATION_NO_PREVIOUS_OPS: &str = "A366616374696F6E6664656C65746566736368656D617849636861745F30303230633635353637616533376566656132393365333461396337643133663866326266323364626463336235633762396162343632393331313163343866633738626776657273696F6E01";
```
Comment (sandreae) [inline aquadoggo/src/graphql/client/mutation.rs:131]: I'm not really sure where these should live :raised_eyebrow:

Comment (adzialocha) [reply aquadoggo/src/graphql/client/mutation.rs:131]: If they are so special then its fine where they are. Slightly unrelated: One could also use the cbor macro to generate CBOR data from some JSON-like interface. Then its easy to change the data + you can see what is inside of it.

Comment (sandreae) [reply aquadoggo/src/graphql/client/mutation.rs:131]: > One could also use the cbor macro

Oh yeh, that'd be better :+1:

File: aquadoggo/src/graphql/client/mutation.rs [hunk 2]
```diff
@@ -192,4 +212,385 @@ mod tests {
...
+                1,
+                Some(DEFAULT_HASH.parse().unwrap()),
+                Some(DEFAULT_HASH.parse().unwrap()),
+                Some(Operation::from(&OperationEncoded::new(OPERATION_ENCODED).unwrap()))
+,
+            key_pair(DEFAULT_PRIVATE_KEY)
+        ),
+        OPERATION_ENCODED,
+        "Could not decode payload hash DecodeError"
+    )]
```
Comment (cafca) [inline aquadoggo/src/graphql/client/mutation.rs:360]: Note: Example for error handling discussion

File: aquadoggo/src/graphql/client/mutation.rs (hunk 1 above)
Comment (sandreae) [reply aquadoggo/src/graphql/client/mutation.rs:131]: I had a go at this and got into trouble defining them as `const` (cos of having to make fn calls).... Maybe not worth holding up the PR for slightly nicer const strings :rofl:

Comment (adzialocha) [reply aquadoggo/src/graphql/client/mutation.rs:131]: Sure, its fine as it is :smiley_cat:

File: aquadoggo/src/graphql/client/mutation.rs [hunk 3]
```diff
@@ -198,4 +219,385 @@ mod tests {
...
+              "query": publish_entry_request.query,
+              "variables": publish_entry_request.variables
+            }
+            ))
+            .send()
+            .await;
+
+        let response = response.json::<serde_json::Value>().await;
+
+        // TODO: I think we'd like a nicer error message here.
```
Comment (adzialocha) [inline aquadoggo/src/graphql/client/mutation.rs:598]: I'd link to this issue here: https://github.com/p2panda/aquadoggo/issues/159

**Review by adzialocha — APPROVED** [review]
Thank you for finally taking care of this. Looks ready to go to me!

**Pull request merged**